ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,160 @@
+
+  let select ~__context ~self =
+    debug "%s: %s" __FUNCTION__ (Ref.string_of self) ;
+    let drv = Db.Driver_variant.get_driver ~__context ~self in
+    let d = Db.Host_driver.get_record ~__context ~self:drv in
+    let v = Db.Driver_variant.get_record ~__context ~self in
+    let stdout =
+      drivertool ["select"; d.API.host_driver_name; v.API.driver_variant_name]
+    in
+    info "%s: %s" __FUNCTION__ stdout ;
+    Db.Host_driver.set_selected_variant ~__context ~self:drv ~value:self
+end
+
+let create ~__context ~host ~name ~friendly_name ~_type ~description ~info:inf
+    ~active_variant ~selected_variant =
+  D.info "%s: %s" __FUNCTION__ name ;
+  let ref = Ref.make () in
+  let uuid = Uuidx.to_string (Uuidx.make ()) in
+  Db.Host_driver.create ~__context ~ref ~uuid ~host ~name ~friendly_name ~_type
+    ~description ~info:inf ~active_variant ~selected_variant ;
+  ref
+
+(** destroy driver and recursively its variants, too *)
+let destroy ~__context ~self =
+  D.debug "Destroying driver %s" (Ref.string_of self) ;
+  let variants = Db.Host_driver.get_variants ~__context ~self in
+  variants |> List.iter (fun self -> Variant.destroy ~__context ~self) ;
+  Db.Host_driver.destroy ~__context ~self
+
+(** Runs on the host where the driver is installed *)
+let select ~__context ~self ~variant =
+  let drv = Ref.string_of self in
+  let var = Ref.string_of variant in
+  let variants = Db.Host_driver.get_variants ~__context ~self in
+  match List.mem variant variants with
+  | true ->
+      D.debug "%s selecting driver %s variant %s" __FUNCTION__ drv var ;
+      let d = Db.Host_driver.get_record ~__context ~self in
+      let v = Db.Driver_variant.get_record ~__context ~self:variant in
+      let stdout =
+        drivertool ["select"; d.API.host_driver_name; v.API.driver_variant_name]
+      in
+      info "%s: %s" __FUNCTION__ stdout ;
+      Db.Host_driver.set_selected_variant ~__context ~self ~value:variant
+  | false ->
+      error "%s variant %s does not belong to driver %s" __FUNCTION__ var drv ;
+      invalid_value "variant" var
+
+(** Runs on the host where the driver is installed *)
+let deselect ~__context ~self =
+  D.debug "%s  driver %s" __FUNCTION__ (Ref.string_of self) ;
+  let d = Db.Host_driver.get_record ~__context ~self in
+  let stdout = drivertool ["deselect"; d.API.host_driver_name] in
+  info "%s: %s" __FUNCTION__ stdout ;
+  Db.Host_driver.set_active_variant ~__context ~self ~value:Ref.null ;
+  Db.Host_driver.set_selected_variant ~__context ~self ~value:Ref.null
+
+(** remove all host driver entries for this host *)
+let reset ~__context ~host =
+  D.debug "%s" __FUNCTION__ ;
+  let open Xapi_database.Db_filter_types in
+  let expr = Eq (Field "host", Literal (Ref.string_of host)) in
+  let drivers = Db.Host_driver.get_refs_where ~__context ~expr in
+  drivers |> List.iter (fun self -> destroy ~__context ~self)
+
+(** Runs on [host] *)
+let scan ~__context ~host =
+  T.Mock.install () ;
```

Comment:
If you always install the mock tool here, then it will be hard to test the real one.

Reply:
Once the real one comes along, we will have changes anyway and I will remove this one.